src/compiler/semantic/mod.rs

```diff
@@ -0,0 +1,342 @@
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Display},
+};
+
+use num_bigint::BigInt;
+
+use itertools::Itertools;
+
+use crate::{
+    compiler::semantic::analyser::Analyser,
+    parser::ast::{
+        expression::Expression, statement::Statement, tl::TLDecl, DebugSymRef, Identifier,
+    },
+};
+
+/// Category of a symbol
+#[derive(Clone, PartialEq, Debug)]
+enum SymbolCategory {
+    Machine,
+    State,
+
+    Signal,
+    WGVar,
+
+    InputSignal,
+    OutputSignal,
+    InoutSignal,
+
+    InputWGVar,
+    OutputWGVar,
+    InoutWGVar,
+}
+
+/// Category of a scope
+#[derive(Clone, Debug)]
+pub enum ScopeCategory {
+    Global,
+    Machine,
+    State,
+}
+
+/// Information about a symbol
+#[derive(Clone, Debug)]
+pub struct SymTableEntry {
+    definition_ref: DebugSymRef,
+    category: SymbolCategory,
+    ty: Option<String>,
```

Comment:
What is the `ty`?

Reply:
Type, it is a reserved word so usually in rust I saw ty used but I will add comment